taichi/runtime/llvm/runtime.cpp

```diff
@@ -637,6 +638,16 @@ void taichi_assert(Context *context, i32 test, const char *msg) {
   taichi_assert_runtime(context->runtime, test, msg);
 }
 
+const std::size_t ASSERT_MSG_BUFFER_SIZE = 2048;
+char assert_msg_buffer[ASSERT_MSG_BUFFER_SIZE];
+char *get_assert_msg(const char *format, ...) {
+  std::va_list args;
+  va_start(args, format);
+  std::vsnprintf(assert_msg_buffer, ASSERT_MSG_BUFFER_SIZE, format, args);
```

Comment:
Calling `std::vsnprintf` here may lead to compilation/linking issues when compile/loading the runtime LLVM bitcode. Please consider passing it in as a function pointer, just like `Runtime::host_printf`. You will have to add an argument `Runtime *` to `get_assert_msg`. Use `get_runtime()` to get the pointer to the `LLVMRuntime` struct at codegen. To learn about how to set `LLVMRuntime::host_vsnprintf`, just search for `host_printf` and see how this function is treated in a similar manner.